bingo.js

```diff
@@ -602,9 +675,119 @@ function setSquareColor(square, colorClass)
 {
 	ALL_COLOURS.forEach(c => square.removeClass(c));
 	square.addClass(colorClass);
+	updateCurrentSheetProgress();
 }
 
-function copySeedToClipboard(id, event)
+function loadBingoProgress(sheetToLoad)
+{
+	/* 
+	 * Expected input for sheetToLoad is [1,3,2,5,..]
+	 * where index of array is numeric value of square id
+	 * and cell value is index of ALL_COLOURS 
+	 */
+	if (typeof sheetToLoad !== 'undefined' && sheetToLoad.length) {
+		forEachSquare((i, square) => {
+			$.each(ALL_COLOURS, function(index, colorClass){
+	       square.removeClass(colorClass);
+	    });
+			square.addClass(ALL_COLOURS[sheetToLoad[i]]);
+		});
+	}
+}
+
+var sheetProgressDebounce = null;
+function updateCurrentSheetProgress()
+{
+	if(sheetProgressDebounce) clearTimeout(sheetProgressDebounce);
+	sheetProgressDebounce = setTimeout(function() {
+			updateCurrentSheetProgressDebounced()
+	}, 100);
+}
+
+function updateCurrentSheetProgressDebounced()
+{
+	if (! isLocalStorageIsAvailable()) {
+		return;
+	}
+
+	processSheetAndStoreAs(SHEET_PROGRESS_KEY + generateNewUrlParam().replace(/-\d-\d/, ''));
+}
+
+function processSheetAndStoreAs(localStorageKey)
+{
+	var squares = generateSquareProgress();
+	result = {
+		squares: squares,
+		timestamp: Date.now(),
+		haveMarkedSquares: squares.reduce(function(a,b){return !!a | !!b})
+	};
+	
+	// save only if current sheet have marked squares or already in storage
+	if (result.haveMarkedSquares || localStorage.getItem(localStorageKey)) {
+		pushNewLocalSetting(localStorageKey, JSON.stringify(result));
+	}
+	// update &progress param is such exists
+	if (gup(URL_PARAM_PROGRESS)) {
+		pushNewUrl(generateNewUrlParamWithProgress(squares));
+	}
+	checkAndRemoveOldSaves();
+}
+
+function checkAndRemoveOldSaves()
+{
+	var saves = getSaves();
+	if (saves.length > SHEET_PROGRESS_SAVE_LIMIT) {
+		var oldestSave = saves.slice(-1).pop();
+		if (oldestSave) {
+			localStorage.removeItem(oldestSave.key);
+		}
+	}
+}
+
+function getSaves()
+{
+	var saves = [];
+	for (i = 0; i < localStorage.length; i++) {
+		key = localStorage.key(i);
+		if (key.slice(0, SHEET_PROGRESS_KEY.length) === SHEET_PROGRESS_KEY) {
+			saves.push({
+				key: key,
+				data: JSON.parse(localStorage.getItem(key))
+			});
+		}
+	}
+
+	saves.sort(function(a, b) {
+		// Compare the 2 dates
+		if (a.data.timestamp < b.data.timestamp) return 1;
+		if (a.data.timestamp > b.data.timestamp) return -1;
+		return 0;
+	});
+	return saves;
+}
+
+function generateSquareProgress()
+{
+	var squares = []
+	forEachSquare((i, square) => {
+		squares.push(ALL_COLOURS.indexOf(square.attr('class') || ''));
```

Comment:
If I understand correctly, hard-coded empty string `''` here represents the grey squares. This would conflict with changes in commit f98eb7219cb46a3c3eb6f9634ca4f611438eb8d1 from #117.

Reply:
Using the whole value of the `class` attribute will break if some other non-colour related classes are added to squares in the future. Using function `getColourClass` would be more future-proof.

Reply:
> Using function `getColourClass` would be more future-proof.

Actually, `getColourClass` uses `Array.prototype.find`. Using [`findIndex`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex) would make more sense here.

Reply:
Should I 'cherry-pick' `const DEFAULT_SQUARE_CLASS_NAME = "greysquare";` to my branch and use it?

Reply:
Maybe I've missed the point, but I don't feel like this:
<img width="734" alt="Снимок экрана 2021-04-11 в 19 09 32" src="https://user-images.githubusercontent.com/16069758/114311987-ae7e9e80-9af9-11eb-84ab-52281a086780.png">
is more readable than this:
<img width="520" alt="Снимок экрана 2021-04-11 в 19 10 53" src="https://user-images.githubusercontent.com/16069758/114311990-b0486200-9af9-11eb-90b9-5afa92be218f.png">

Redoing `getColourClass` function to use `findIndex` would be an option, but it will add additional `if`'s to have an index of `DEFAULT_SQUARE_CLASS_NAME` as a default

Reply:
I meant using `hasClass` in function `generateSquareProgress` like this:

```js
 	forEachSquare((i, square) => {
		squares.push(ALL_COLOURS.findIndex(c => square.hasClass(c)));
 	});
```

and no changes in `getColourClass`.

Reply:
As for conflicts, please do not use cherry-pick. Cherry-pick creates a new commit with the same changes, which means that there would be two commits with the same changes in history, which is bad.

I recommend rebasing on top of `dev` after Josh merges either of the pull requests.

Reply:
```
forEachSquare((i, square) => {
    squares.push(ALL_COLOURS.findIndex(c => square.hasClass(c)));
});
```

This would work, but in my branch there is no class for default color, meaning there would be a lot of `-1` in `progress` by default. I'll see what I can do for default value, to keep it as much one-liner as possible :)

Reply:
```JavaScript
function getColourClassIndex(square)
{
	return ALL_COLOURS.findIndex(c => square.hasClass(c));
}

forEachSquare((i, square) => {
	var classIndex = getColourClassIndex(square);
	squares.push(classIndex !== -1 ? classIndex : 0);
});
```

I've added separate function for getting just index (since there was similar usage in other method), and added default index as `0`, so I think this shouldn't have conflict now with changes to `DEFAULT_SQUARE_CLASS_NAME`